### DISCUSSION TITLE
[SoftBody] Burst and Cracks Along Seams for Sphere and Box

### DISCUSSION THREAD

**Discussion by alienself** [jrouwe/JoltPhysics.js#208] [Q&A] [answered]
### Description

I am encountering an issue where meshes burst or deform unexpectedly upon collision with the floor. Specifically:

- **Sphere:** A crack forms along the seam from the north pole to the south pole upon impact.
- **Box:** Each face appears to disconnect from the others as if the seams were not glued together.

This behavior occurs as soon as the mesh collides with the floor, and the meshes appear to "burst" like a balloon. I have tried adjusting the pressure values but this has not resolved the issue.

### Steps to Reproduce
1. Simulate a sphere or box mesh with softbody.
2. Set up a collision with a floor or plane.
3. Observe the deformation upon impact.

### Expected Behavior
- Sphere and box should maintain their structural integrity without cracking or disconnection at the seams upon collision.

### Actual Behavior
- The sphere forms a crack from the north to south pole upon collision.
- The box faces disconnect as if the seams are not properly glued.

### Additional Information
I have experimented with different pressure values, but none seem to have a significant effect on this issue. Any guidance or suggestions would be appreciated.

<img width="826" alt="374425193-b5507578-dc09-4ee5-a829-6fa620650b85" src="https://github.com/user-attachments/assets/88a73ce2-75af-47f0-85ea-197c08710e98">

<img width="654" alt="374425201-5bdb0341-e178-4eaa-b27c-9c1728a8a3b4" src="https://github.com/user-attachments/assets/668d59b9-e04b-4263-ab6c-826ce6bdf42e">

Here is the code I use to generate the soft body from an indexed mesh:

```ts
function createFromIndexedMesh(
      vertices: Float32Array, 
      indices: Uint32Array, 
      inBendType: SoftBodySharedSettings_EBendType = Jolt.SoftBodySharedSettings_EBendType_None, 
      inVertexAttributes?: SoftBodySharedSettingsVertexAttributes
  ): SoftBodySharedSettings {
      if (vertices.length % 3 !== 0) {
          throw new Error("Vertices array length must be a multiple of 3.");
      }

      if (indices.length % 3 !== 0) {
          throw new Error("Indices array length must be a multiple of 3.");
      }

      // Create settings
      const sharedSettings: SoftBodySharedSettings = new Jolt.SoftBodySharedSettings;
      const v = new Jolt.SoftBodySharedSettingsVertex;

      // ---

      // Add vertices
      for (let i = 0; i < vertices.length; i += 3) {
          v.mPosition.x = vertices[i];
          v.mPosition.y = vertices[i + 1];
          v.mPosition.z = vertices[i + 2];
          v.mInvMass = 1.0; // Assuming a default inverse mass of 1 for all vertices, can be customized
          sharedSettings.mVertices.push_back(v);
      }
      Jolt.destroy(v);

      console.log("SOFTBODY:vertices", vertices.length )


      // ---

      // Add faces based on the indices
      const f = new Jolt.SoftBodySharedSettingsFace(0, 0, 0, 0);
      for (let i = 0; i < indices.length; i += 3) {
          f.set_mVertex(0, indices[i]);
          f.set_mVertex(1, indices[i + 1]);
          f.set_mVertex(2, indices[i + 2]);
          sharedSettings.AddFace(f);
      }
      Jolt.destroy(f);

      // ---

      // Edge constraints (assuming a simple edge compliance across connected vertices)
      const sEdge = new Jolt.SoftBodySharedSettingsEdge(0, 0, 0);
      sEdge.mCompliance = 0.00001;

      // Loop through the indices to create edges based on connections in the mesh
      const numTriangles = indices.length / 3;
      for (let i = 0; i < numTriangles; ++i) {
          const i0 = indices[i * 3];
          const i1 = indices[i * 3 + 1];
          const i2 = indices[i * 3 + 2];

          // Create edges between vertices in each triangle
          sEdge.set_mVertex(0, i0);
          sEdge.set_mVertex(1, i1);
          sharedSettings.mEdgeConstraints.push_back(sEdge);

          sEdge.set_mVertex(0, i1);
          sEdge.set_mVertex(1, i2);
          sharedSettings.mEdgeConstraints.push_back(sEdge);

          sEdge.set_mVertex(0, i2);
          sEdge.set_mVertex(1, i0);
          sharedSettings.mEdgeConstraints.push_back(sEdge);
      }
      Jolt.destroy(sEdge);

      // Calculate edge lengths and volume constraints if applicable
      sharedSettings.CalculateEdgeLengths();



      // ---


      // Create volume constraints if necessary
      // (Implementation would be based on your specific requirements, similar to the cube example)

      if (inVertexAttributes) {
          sharedSettings.CreateConstraints(inVertexAttributes, 1, inBendType);
      } else {
          const defaultAttributes = new Jolt.SoftBodySharedSettingsVertexAttributes();
          defaultAttributes.mCompliance = 1.0e-5;
          defaultAttributes.mShearCompliance = 1.0e-5;
          defaultAttributes.mBendCompliance = 1.0e-5;
          sharedSettings.CreateConstraints(defaultAttributes, 1, inBendType);
          Jolt.destroy(defaultAttributes);
      }

      // Optimize the settings
      sharedSettings.Optimize();


      return sharedSettings;
  }
  ```

**Comment by jrouwe** [comment, marked as answer]
Are you sure that your sphere / box mesh is actually closed? E.g. a box is usually modelled like this:

![image](https://github.com/user-attachments/assets/c7ac444f-175b-4806-b21e-709f58bfa15d)

where vertex 'v' is actually 3 different vertices because you need 3 normals (in green) to get correct lighting.

For spheres it's the same thing. There's usually a seam where the UV coordinate goes from 1 back to 0 and at that point there are 2 vertices with the same position, same normal but different UV coordinates.

You need to ensure that for the simulation mesh there is only 1 vertex per position.

**Reply by alienself** [reply]
Apologies for the late reply, your advice was super helpful. I recently managed to get it working correctly by welding the faces before creating the soft body. Thank you very much for your assistance.
